File: apps/java/src/main/java/chapter08/MostFrequentImputer.java
package chapter08;

import chapter02.Row;
import chapter02.Table;
import java.util.LinkedHashMap;
import java.util.Map;

/** 文字列の列の欠損値を、訓練データで最も多い値（最頻値）で補完する前処理。 */
public record MostFrequentImputer(String column) implements Transformer {
  @Override
  public FittedTransformer fit(Table x) {
    Map<String, Integer> counts = new LinkedHashMap<>();
    x.rows().stream()
        .filter(row -> !row.isMissing(column))
        .forEach(row -> counts.merge(row.text(column), 1, Integer::sum));
    String mostFrequent = null;
    int bestCount = 0;
    for (Map.Entry<String, Integer> entry : counts.entrySet()) {
      if (entry.getValue() > bestCount) {
        mostFrequent = entry.getKey();
        bestCount = entry.getValue();
      }
    }
    return new FittedMostFrequentImputer(column, mostFrequent);
  }

  /** fit で求めた最頻値を持ち、欠損値を補完する。 */
  public record FittedMostFrequentImputer(String column, String mostFrequent)
      implements FittedTransformer {
    @Override
    public Table transform(Table x) {
      return new Table(x.columns(), x.rows().stream().map(this::fill).toList());
    }

    private Row fill(Row row) {
      return row.isMissing(column) ? Rows.with(row, column, mostFrequent) : row;
    }
  }
}

File: apps/csharp/src/MachineLearning/Chapter08/MostFrequentImputer.cs
namespace MachineLearning.Chapter08;

using MachineLearning.Chapter02;

/// <summary>文字列の列の欠損値を、訓練データで最も多い値（最頻値）で補完する前処理。</summary>
public sealed record MostFrequentImputer(string Column) : ITransformer
{
    public IFittedTransformer Fit(Table x)
    {
        ArgumentNullException.ThrowIfNull(x);
        var counts = new Dictionary<string, int>(StringComparer.Ordinal);
        foreach (var row in x.Rows.Where(row => !row.IsMissing(this.Column)))
        {
            counts[row.Text(this.Column)] = counts.GetValueOrDefault(row.Text(this.Column)) + 1;
        }

        var mostFrequent = counts.Aggregate((best, next) => next.Value > best.Value ? next : best).Key;
        return new Fitted(this.Column, mostFrequent);
    }

    /// <summary>Fit で求めた最頻値を持ち、欠損値を補完する。</summary>
    public sealed record Fitted(string Column, string MostFrequent) : IFittedTransformer
    {
        public Table Transform(Table x)
        {
            ArgumentNullException.ThrowIfNull(x);
            return new Table(x.Columns, [.. x.Rows.Select(row => this.Fill(x.Columns, row))]);
        }

        private Row Fill(IReadOnlyList<string> columns, Row row) =>
            row.IsMissing(this.Column) ? Rows.With(columns, row, this.Column, this.MostFrequent) : row;
    }
}
